packages/angular-query-experimental/src/createBaseQuery.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  DestroyRef,
+  assertInInjectionContext,
+  computed,
+  effect,
+  inject,
+  signal,
+} from '@angular/core'
+import { notifyManager } from '@tanstack/query-core'
+import { QueryClientService } from './QueryClientService'
+import type { QueryClient, QueryKey, QueryObserver } from '@tanstack/query-core'
+import type { CreateBaseQueryOptions, CreateBaseQueryResult } from './types'
+
+/**
+ * Base implementation for `createQuery` and `createInfiniteQuery`.
+ * @internal
+ */
+export function createBaseQuery<
+  TQueryFnData,
+  TError,
+  TData,
+  TQueryData,
+  TQueryKey extends QueryKey,
+>(
+  options: () => CreateBaseQueryOptions<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  >,
+  Observer: typeof QueryObserver,
+  queryClient?: QueryClient,
+): CreateBaseQueryResult<TData, TError> {
+  assertInInjectionContext(createBaseQuery)
+  /** Load query client */
+  if (!queryClient) {
+    queryClient = inject(QueryClientService).useQueryClient()
+  }
+
+  /** Creates a signal that has the default options applied */
+  const defaultedOptionsSignal = computed(() => {
+    const defaultedOptions = queryClient!.defaultQueryOptions(options())
+    defaultedOptions._optimisticResults = 'optimistic'
+    return defaultedOptions
+  })
+
+  /** Creates the observer */
+  const observer = new Observer<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  >(queryClient, defaultedOptionsSignal())
+
+  effect(() => {
+    // Do not notify on updates because of changes in the options because
+    // these changes should already be reflected in the optimistic result.
+    observer.setOptions(defaultedOptionsSignal(), { listeners: false })
+  })
+
+  const result = signal(observer.getCurrentResult())
+  const unsubscribe = observer.subscribe(
+    notifyManager.batchCalls((val) => {
+      result.set(val)
```

Comment:
In my case I need to call result.set inside injector.get(NgZone).run otherwise change detected where not triggered for some case. Not sure it is required

Reply:
Do you maybe have a sandbox demonstrating this? I don't think I understand the issue.